src/error/tests.rs

```
@@ -0,0 +1,70 @@
+use super::*;
+
+mod longest_match {
+    use super::*;
+    use crate::{
+        combinator::{alt, eof, terminated},
+        Located,
+    };
+
+    type Input<'a> = Located<&'a str>;
+    type Error<'a> = LongestMatch<Input<'a>, ContextError<&'static str>>;
+    type PResult<'a, O> = crate::error::PResult<O, Error<'a>>;
+
+    fn tag<'a>(t: &'static str) -> impl Parser<Input<'a>, &'a str, Error<'a>> {
+        move |input: &mut Input<'a>| {
+            let mut i = input.clone();
+            for mut c in t.chars() {
+                c.parse_next(input)?;
+            }
+            Ok(i.next_slice(t.len()))
+        }
+    }
+
+    fn syms<'a>(t: &'static str) -> impl Parser<Input<'a>, &'a str, Error<'a>> {
+        terminated(tag(t), eof).context(t)
+    }
+
+    fn expect_err<T: core::fmt::Debug>(res: PResult<'_, T>) -> Vec<&'static str> {
+        let err = res.unwrap_err();
+        let err = err.into_inner().unwrap();
+        let err = err.into_inner();
+        err.context().copied().collect::<Vec<_>>()
+    }
+
+    #[test]
+    fn single_longest_match_first_in_alt() {
+        fn parser<'a>(input: &mut Input<'a>) -> PResult<'a, &'a str> {
+            alt((syms("abcd"), syms("abc"), syms("ab"))).parse_next(input)
+        }
+        let ctx = expect_err(parser(&mut Input::new("abcde")));
+        assert_eq!(ctx, vec!["abcd"]);
+    }
+
+    #[test]
+    fn multi_longest_match() {
+        fn parser<'a>(input: &mut Input<'a>) -> PResult<'a, &'a str> {
+            alt((syms("abcd"), syms("abc"), syms("def"), syms("defg"))).parse_next(input)
+        }
+
+        let ctx = expect_err(parser(&mut Input::new("abd")));
+        assert_eq!(ctx, vec!["abcd", "abc"]);
+
+        let ctx = expect_err(parser(&mut Input::new("deg")));
+        assert_eq!(ctx, vec!["def", "defg"]);
+    }
+
+    #[test]
+    #[ignore]
```

Why is this test ignored?

---

The test is failing because. That one shows the actual problem we face when using `Ord` on `Checkpoint`. 

For context:

> Unfortunately I think `Ord` is not the right trait. I also added some tests to see if it works as expected.
The problem becomes obvious when having partial input. In that case we have a parser that has consumed all input, but did produce an error and a second parser that did fail early. In that case we are comparing `""` with `"rest"`. Apparently `"rest" > ""`, which makes us report the wrong context.
>
> In case of `&'a str` as input [`SliceOrd`](https://doc.rust-lang.org/std/primitive.slice.html#impl-Ord-for-%5BT%5D) will be used. Stepping through with the debugger it looks like [this implementation is used](https://doc.rust-lang.org/src/core/slice/cmp.rs.html#179).

---

Introducing AsOrd did fix the test.